activation/nipost.go

```diff
@@ -83,17 +83,20 @@ func (nb *NIPostBuilder) persistState() {
 
 // NIPostBuilder holds the required state and dependencies to create Non-Interactive Proofs of Space-Time (NIPost).
 type NIPostBuilder struct {
```

Comment:
The `NIPostBuilder` now contains three things that should probably be one:
- a `nipostClient` that returns some information about POST initialization (status, opts and commitment ATX). The 2 latter are required to be able to verify a POST proof
- a `postClient` to _generate_ POST proof
- a `nipostValidator` to _verify_ POST proof

Additionally, it must handle connections and disconnections of the `postClient`, which doesn't feel its concern. Ideally, it should call `.Proof()` and get a `VALID|INVALID|TRY AGAIN` response.

I think it should be possible to merge these 3 into a single interface covering what really the builder is interested in:

```go
type postClient interface {
    GenerateProof(ctx context.Context, challenge []byte) (*types.Post, *types.PostMetadata, error)
    VerifyProof(ctx context.Context, proof *types.Post, proofMetadata *types.PostMetadata) error // The client should probably be able to figure out the commitmentATX for on its own
}
```

---

IMHO the following:
```go
if s := nb.nipostClient.Status(); s.State != PostSetupStateComplete {
    return nil, errors.New("post setup not complete")
}
```

- doesn't make much sense as a check "if the user completed POST initialization" as he must had completed it already because the `nipostChallenge` argument either contains the initial POST proof or is for a subsequent ATX
- isn't the right place as a check "if the user still has the POST data" - this check should be in _generate proof_ routine, most likely on the post-service side, but could be handled on the node side for now too.

WDYT?

Reply:
> - a `nipostClient` that returns some information about POST initialization (status, opts and commitment ATX). The 2 latter are required to be able to verify a POST proof

As the go-doc for this interface explains: it is temporary because and will be merged with the `postClient` interface in the future.

The reason for it being two interfaces is purely a technical one: `nipostClient` is implemented by a different service than `postClient`; merging these now would need an adapter service facilitating calls to `grpcserver.PostClient` and `activation.PostSetupManager` just to later drop this adapter again when the full functionality is available via the `grpcserver.PostClient`.

---

> Additionally, it must handle connections and disconnections of the postClient, which doesn't feel its concern.

It doesn't really, it is just signalled when a new post service connects or an exiting connection is dropped - and handle these events accordingly. Alternatively there could be one `atxBuilder` that is initialised and started for every post service that connects, but this needs to "handle connections" as well, because it needs to be able to be started & stopped at any time when a service connects or disconnects.

It depends if we want `atxBuilder` to be a singleton that builds ATXs for every post service/identity or have a dedicated `atxBuilder` for every post service/identity? Right now the first approach is taken, similar to the implementation for `hare3`: https://github.com/spacemeshos/go-spacemesh/pull/5126/

The second approach would make the `atxBuilder` itself simpler, but could cost performance; for instance if there is one `atxBuilder` per identity every db query needs to be done once for every identity while the one `atxBuilder` approach would allow to combine queries into a single one or remove redundancy.

---

> I think it should be possible to merge these 3 into a single interface covering what really the builder is interested in:

I am not sure about this. I think we need at least 2 interfaces because proof generation and verification are different concerns in my opinion:

-  Proof generation is done by the post service, while proof verification is done by the node - so both methods would be implemented differently.
- Verification must be possible even if no post service connects and it makes no sense to verify a proof for every post service that connects, i.e. there is always exactly one provider for verification, while we can have any number of proof generators (even zero).

---

Regarding the `Status` call: I assume this was added as a sanity check in case something changes between the last call to `BuildNiPoST` and the current one: e.g. data corruption, files deleted, etc.

We can probably drop this call and replace it with a check if at least one post service is connected and ready to generate a proof (since we are already within the cycle gap when this call happens).

I'm thinking of adding a function to the gRPC API that returns the information in the `postdata_metadata.json` file (without generating a proof) with an additional status: "I am (not) ready to generate a proof" - where not being ready could mean that it is already generating a proof, or data is corrupted/missing, etc.

WDYT?

Reply:
> > Additionally, it must handle connections and disconnections of the postClient, which doesn't feel its concern.
> 
> It doesn't really, it is just signalled when a new post service connects or an exiting connection is dropped - and handle these events accordingly. Alternatively there could be one atxBuilder that is initialised and started for every post service that connects, but this needs to "handle connections" as well, because it needs to be able to be started & stopped at any time when a service connects or disconnects.

What I'm trying to say is that the "remoteness" of the post service is not a concern of the `nipostBuilder`. It just really wants to "generate a proof". The fact that there is something that "connects" and that there is some "grpc" call in between is not its concern at all.

>It depends if we want atxBuilder to be a singleton that builds ATXs for every post service/identity or have a dedicated atxBuilder for every post service/identity? Right now the first approach is taken, similar to the implementation for hare3: https://github.com/spacemeshos/go-spacemesh/pull/5126

It has nothing to do with it. Either there will be some "ATX builder manager" orchestrating many AtxBuilders or AtxBuilder orchestrating many IDs. It doesn't matter for this conversation. (Btw, I think error handling will be simpler if we keep 1 AtxBuilder per ID).

---

>> I think it should be possible to merge these 3 into a single interface covering what really the builder is interested in:
>
>I am not sure about this. I think we need at least 2 interfaces because proof generation and verification are different concerns in my opinion:
> - Proof generation is done by the post service, while proof verification is done by the node - so both methods would be implemented differently.
> - Verification must be possible even if no post service connects and it makes no sense to verify a proof for every post service that connects, i.e. there is always exactly one provider for verification, while we can have any number of proof generators (even zero).

The fact that one is talking with some post-service and the other is not is an implementation detail that is not a concern of the `nipostBuilder` and could change at any moment without impacting the builder. 

:bulb: Actually, we could argue that verification could be removed entirely now as the proof is already verified in the post-service. It would simplify the builder even more. Wdyt?

---

> Regarding the Status call: I assume this was added as a sanity check in case something changes between the last call to BuildNiPoST and the current one: e.g. data corruption, files deleted, etc.

Could be. As such I think it makes little sense to keep it.

> We can probably drop this call and replace it with a check if at least one post service is connected and ready to generate a proof (since we are already within the cycle gap when this call happens).

Wdym? There will be exactly 1 post-service per ID. If it's not connected, we retry until the publish epoch ends.

> I'm thinking of adding a function to the gRPC API that returns the information in the postdata_metadata.json file (without generating a proof) with an additional status: "I am (not) ready to generate a proof" - where not being ready could mean that it is already generating a proof, or data is corrupted/missing, etc.

Not part of PR - let's discuss it offline

Reply:
To summarize the discussion with @poszu:

- We aim to remove `PostClient` as well the `Connected`/`Disconnected` methods signalling when a PoST service connects.
- Services that are dependent on `PostClient` at the moment will instead depend on `grpcserver::PostService`. `grpcserver::PostService` will expose methods like `Proof` with the intended `NodeId` passed as argument, so that it can forward the command to the correct post service.